### ISSUE TITLE
add early stopping?

### ISSUE THREAD

**Issue by nick-youngblut** [RasmussenLab/vamb#57]
Given that 500 epochs can take a while with many 100's of 1000's of contigs, it could be useful to implement early stopping, which isn't too hard to add in pytorch. For example:

```
class EarlyStopping:
    """
    Early stops the training if validation loss doesn't improve after a given patience.
    """
    def __init__(self, patience=10, verbose=False, delta=0,
                 path='checkpoint.pt', trace_func=print):
        """
        Args:
            patience (int): How long to wait after last time validation loss improved.
                            Default: 7
            verbose (bool): If True, prints a message for each validation loss improvement.
                            Default: False
            delta (float): Minimum change in the monitored quantity to qualify as an improvement.
                            Default: 0
            path (str): Path for the checkpoint to be saved to.
                            Default: 'checkpoint.pt'
            trace_func (function): trace print function.
                            Default: print
        """
        self.patience = patience
        self.verbose = verbose
        self.counter = 0
        self.best_score = None
        self.early_stop = False
        self.val_loss_min = np.Inf
        self.delta = delta
        self.path = path
        self.trace_func = trace_func

    def __call__(self, val_loss, model):
        score = -val_loss

        if self.best_score is None:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
        elif score < self.best_score + self.delta:
            self.counter += 1
            self.trace_func(f'  Early-Stop counter: {self.counter} out of {self.patience}')
            if self.counter >= self.patience:
                self.early_stop = True
        else:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
            self.counter = 0
        return self.early_stop

    def save_checkpoint(self, val_loss, model):
        """Saves model when validation loss decrease."""
        if self.verbose:
            msg = 'Validation loss decreased ({0:.6f} --> {1:.6f}).  Saving model ...'
            self.trace_func(msg.format(self.val_loss_min, self.val_loss))
        torch.save(model.state_dict(), self.path)
        self.val_loss_min = val_loss
```

**Comment by simonrasmu** [comment]
Hi Nick, 

Was thinking that Jakob would chip in first. My thoughts on this is that it could be done, but we just really have to be careful as this could have an implication for the clustering step that use the latent representation. So really we would need to benchmark how it performs. However, we will certainly put it on the to-do list!

**Comment by nick-youngblut** [comment]
I've noticed that for a lot of the VAMB runs that I've done, the loss doesn't change much (ie., <0.001 in changes) beyond  50-200 epochs, so early stopping could prevent a lot of unnecessary epochs

**Comment by simonrasmu** [comment]
Yes indeed it is the same for us - we should certainly benchmark it. You could try to run the same data several times with e.g. 50, 100, 200, 300, 400, 500 epochs and compare. We can try that as well on our benchmark datasets and see how it performs

**Comment by jakobnissen** [comment]
Unfortunately, I seem to remember that the latent representation continues to change even after loss appears to flatten out. So we can't simply look at the loss to see how the clustering will perform.
We did actually benchmark it - Supplementary figure 33 in our paper. It seems to work fine with 250 epochs, but that's on the smaller benchmark datasets.

I've changed the default number of epochs to 300.
